x-pack/metricbeat/module/aws/awshealth/awshealth.go

```diff
@@ -0,0 +1,439 @@
+// identified by a Event ARN. The function returns a slice of mb.Event structs
+// containing the summarized event info.
+func (m *MetricSet) getEventsSummary(
+	ctx context.Context,
+	awsHealth *health.Client,
+) []mb.Event {
+	var events []mb.Event
+	eventFilter := types.EventFilter{
+		EventStatusCodes: []types.EventStatusCode{
+			types.EventStatusCodeUpcoming,
+			types.EventStatusCodeOpen,
+		},
+	}
+
+	var (
+		nextTokenString string
+		eventOutput     *health.DescribeEventsOutput
+		err             error
+		wg              sync.WaitGroup
+	)
+	errCh := make(chan error, maxResults)
+
+	// Create buffered channel to receive event with event description and affected entity details
+	c := make(chan HealthDetails, maxResults)
+
+	for {
+		// When invoking the DescribeEvents for the first time, there must not exist any NextToken.
+		// Upon calling the DescribeEvents API, the next token will be returned if there are additional records available for querying.
+		// If there are no more records to fetch, the next token will be empty.
+		if nextTokenString == "" {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+				},
+			)
+		} else {
+			eventOutput, err = awsHealth.DescribeEvents(ctx,
+				&health.DescribeEventsInput{
+					Filter:     &eventFilter,
+					MaxResults: &maxResults,
+					NextToken:  &nextTokenString,
+				},
+			)
+		}
+		if err != nil {
+			err = fmt.Errorf("[AWS Health] DescribeEvents failed with : %w", err)
+			m.Logger().Error(err.Error())
+			return nil
+		}
+		ets := eventOutput.Events
+		select {
+		case <-ctx.Done():
+			// Context cancelled, handle graceful termination
+			close(c)
+			return nil
+		default:
+			// Context not cancelled, proceed with the function
+		}
+
+		for i := range ets {
+			m.Logger().Debugf("[AWS Health] [Fetch DescribeEventDetails] Event ARN : %s", getValueOrDefault(ets[i].Arn, ""))
+			// Increment the WaitGroup counter
+			wg.Add(1)
+			go func(et types.Event) {
+				defer wg.Done()
+				err := m.getDescribeEventDetails(ctx, awsHealth, et, c)
+				if err != nil {
+					errCh <- err
+				}
+			}(ets[i])
+		}
+
+		for i := 0; i < len(ets); i++ {
+			select {
+			case <-ctx.Done():
+				// Context cancelled, handle graceful termination
+				m.Logger().Debug("Context cancelled. Exiting gracefully.")
+				close(c)
+				return nil
+			case err := <-errCh:
+				// Handle errors received from goroutines
+				m.Logger().Error(err.Error())
+			case healthDetails, ok := <-c:
+				if !ok {
+					return nil
+				}
+				m.Logger().Debugf("[AWS Health] [DescribeEventDetails] Event ARN : %s, Affected Entities (Pending) : %d, Affected Entities (Resolved): %d, Affected Entities (Others) : %d", *healthDetails.event.Arn, healthDetails.affectedEntityPending, healthDetails.affectedEntityResolved, healthDetails.affectedEntityOthers)
+				events = append(events, createEvents(healthDetails))
+			}
+		}
+		wg.Wait()
```

Comment:
Did you try removing the waitgroup? I do not think you need it given the code that you have.

Try removing `wg.*` and test?

Reply:
Addressed.